src/core/geometry/qgscircularstring.cpp

```diff
@@ -1592,6 +1594,118 @@ double QgsCircularString::segmentLength( QgsVertexId startVertex ) const
+    return -1.0;
+
+  if ( fromVertexNumber == toVertexNumber )
+    return 0.0;
+
+  // Ensure fromVertex < toVertex for simplicity
+  if ( fromVertexNumber > toVertexNumber )
+  {
+    return distanceBetweenVertices( QgsVertexId( 0, 0, toVertexNumber ), QgsVertexId( 0, 0, fromVertexNumber ) );
+  }
+
+  double totalDistance = 0.0;
+  int nPoints = numPoints();
+
+  // Iterate through the arcs, accumulating distance between fromVertex and toVertex
+  for ( int i = 0; i < nPoints - 2; i += 2 )
+  {
+    // Arc segment from i to i+2, with curve point at i+1
+    double x1 = mX.at( i );     // Start point
+    double y1 = mY.at( i );
+    double x2 = mX.at( i + 1 ); // Curve point
+    double y2 = mY.at( i + 1 );
+    double x3 = mX.at( i + 2 ); // End point
+    double y3 = mY.at( i + 2 );
+
+    // Check if both vertices are in this arc segment
+    if ( fromVertexNumber >= i && toVertexNumber <= i + 2 )
+    {
+      if ( fromVertexNumber == i && toVertexNumber == i + 2 )
+      {
+        // Full arc from start to end
+        return QgsGeometryUtilsBase::circleLength( x1, y1, x2, y2, x3, y3 );
+      }
+      else if ( fromVertexNumber == i && toVertexNumber == i + 1 )
+      {
+        // Arc from start point to curve point - use precise calculation
+        double centerX, centerY, radius;
+        QgsGeometryUtils::circleCenterRadius( QgsPoint( x1, y1 ), QgsPoint( x2, y2 ), QgsPoint( x3, y3 ), radius, centerX, centerY );
+        // Calculate arc length from vertex i to vertex i+1
+        int relativeFrom = fromVertexNumber - i;  // Should be 0
+        int relativeTo = toVertexNumber - i;      // Should be 1
+        return QgsGeometryUtilsBase::calculateArcLength( centerX, centerY, radius, x1, y1, x2, y2, x3, y3, relativeFrom, relativeTo );
+      }
+      else if ( fromVertexNumber == i + 1 && toVertexNumber == i + 2 )
+      {
+        // Arc from curve point to end point - use precise calculation
+        double centerX, centerY, radius;
+        QgsGeometryUtils::circleCenterRadius( QgsPoint( x1, y1 ), QgsPoint( x2, y2 ), QgsPoint( x3, y3 ), radius, centerX, centerY );
+        // Calculate arc length from vertex i+1 to vertex i+2
+        int relativeFrom = fromVertexNumber - i;  // Should be 1
+        int relativeTo = toVertexNumber - i;      // Should be 2
+        return QgsGeometryUtilsBase::calculateArcLength( centerX, centerY, radius, x1, y1, x2, y2, x3, y3, relativeFrom, relativeTo );
+      }
+      else if ( fromVertexNumber == i + 1 && toVertexNumber == i + 1 )
+      {
+        return 0.0; // Same point
+      }
+    }
+
+    // Handle cases where vertices span multiple segments
+    bool startInThisSegment = ( fromVertexNumber >= i && fromVertexNumber <= i + 2 );
+    bool endInThisSegment = ( toVertexNumber >= i && toVertexNumber <= i + 2 );
+    bool segmentInRange = ( fromVertexNumber < i && toVertexNumber > i + 2 );
```

Comment:
I'm confused by all this logic -- what does it mean to have vertices spanning multiple segments?

Reply:
Ok, got it :+1: